src/hotspot/share/oops/compressedKlass.hpp

```diff
@@ -62,10 +73,60 @@ class CompressedKlassPointers : public AllStatic {
   static char* reserve_address_space_for_16bit_move(size_t size, bool aslr);
 
-  DEBUG_ONLY(static void assert_is_valid_encoding(address addr, size_t len, address base, int shift);)
+  // Returns the highest address expressable with an unshifted narrow Klass pointer
+  inline static uintptr_t highest_unscaled_address();
+
+  static bool pd_initialize(address addr, size_t len);
+
+#ifdef ASSERT
+  // For sanity checks: Klass range
+  static address _klass_range_start;
+  static address _klass_range_end;
+  // For sanity checks: lowest, highest valid narrow klass ids != null
+  static narrowKlass _lowest_valid_narrow_klass_id;
+  static narrowKlass _highest_valid_narrow_klass_id;
+  static void calc_lowest_highest_narrow_klass_id();
+  static void sanity_check_after_initialization();
+#endif // ASSERT
+
+  template <typename T>
+  static inline void check_init(T var) {
+    assert(var != (T)-1, "Not yet initialized");
+  }
 
 public:
 
+  // Initialization sequence:
+  // 1) Parse arguments. The following arguments take a role:
+  //      - UseCompressedClassPointers
+  //      - UseCompactObjectHeaders
+  //      - Xshare on off dump
+  //      - CompressedClassSpaceSize
+  // 2) call pre_initialize(): depending on UseCompactObjectHeaders, defines the limits of narrow Klass pointer
+  //    geometry (how many bits, the max. possible shift)
+  // 3) .. from here on, narrow_klass_pointer_bits() and max_shift() can be used
+  // 4) call reserve_address_space_for_compressed_classes() either from CDS initialization or, if CDS is off,
+  //    from metaspace initialization. Reserves space for class space + CDS, attempts to reserve such that
+  //    we later can use a "good" encoding scheme. Reservation is highly CPU-specific.
+  // 5) Initialize the narrow Klass encoding scheme by determining encoding base and shift:
+  //   5a) if CDS=on: Calls initialize_for_given_encoding() with the reservation base from step (4) and the
+  //       CDS-intrinsic setting for shift; here, we don't have any freedom to deviate from the base.
+  //   5b) if CDS=off: Calls initialize() - here, we have more freedom and, if we want, can choose an encoding
+  //       base that differs from the reservation base from step (4). That allows us, e.g., to later use
+  //       zero-based encoding.
+  // 6) ... from now on, we can use base() and shift().
+
+  // Called right after argument parsing; defines narrow klass pointer geometry limits
+  static void pre_initialize();
+
+  static bool tiny_classpointer_mode()   { check_init(_tiny_cp); return (_tiny_cp == 1); }
+
+  // The number of bits a narrow Klass pointer has;
+  static int narrow_klass_pointer_bits() { check_init(_narrow_klass_pointer_bits); return _narrow_klass_pointer_bits; }
+
+  // The maximum possible shift; the actual shift employed later can be smaller (see initialize())
+  static int max_shift()                 { check_init(_max_shift); return _max_shift; }
```

Comment:
FWIW, you could get rid of the variable name duplication if we changed this to be something like `{ return read_check_init(_max_shift); }` instead.